SCMOD-15887: Initial implementation

xbreizh opened the pull request:
https://portal.digitalsafe.net/browse/SCMOD-15887

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFapi/job/CAFapi~util-flyway-installer~SCMOD-15887~CI

simranjeetc commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,107 @@
+            description = "Specifies the connection string to the database service. " +
+                    "e.g. postgresql://localhost:3307/"
+    )
+    private String connectionString;
```
Can we not have only fullConnectionString? What is the intent of providing two different flavors here?

xbreizh replied:
As discussed, that would make more sense. I was done so we don't have to modify the calling code from [here](https://github.com/JobService/job-service/blob/af7e3e5465b9018e46832e1ed62253541680442b/job-service-postgres-container/pom.xml#L131) @dermot-hardy what do you think? should we keep it the way it is or simplify to pass only the full url?

dermot-hardy replied:
Our services take the database parameters as individual arguments:
https://github.com/CAFapi/opensuse-base-image#database-creation-script

I would ask the opposite question.  Why do we need `fullConnectionString`?  The code you referenced doesn't pass it.  I certainly agree that the utility shouldn't support two flavors of input without good reason.

simranjeetc commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,107 @@
+            Migrator.migrate(allowDBDeletion, fullConnectionString, connectionString, username, password, dbName);
+            LOGGER.info("Migration completed ...");
+        } catch (final FlywayMigratorException | SQLException e) {
+            LOGGER.error(e.getMessage());
```
While logging errors in catch blocks, use the overloaded method with throwables.

xbreizh replied:
Done

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+    
+    private static void createNewDatabase(final String dbName, final BasicDataSource dbSource) throws SQLException
```
The name mentions it to be creating a new database where in reality it also drops the whole database. This should either be named appropriately or drop can be a separate method.

xbreizh replied:
Done

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+                    .load();
+            flyway.migrate();
+            LOGGER.info("DB update finished.");
+        } catch (FlywayMigratorException e) {
```
final

xbreizh replied:
Done

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+            dbName = fullConnectionUrl.substring(fullConnectionUrl.lastIndexOf('/') + 1);
+            dbSource.setUsername(username);
+            dbSource.setPassword(password);
+            if(checkDBExists(dbSource) || allowDBDeletion) {
```
Won't this drop an existing database even if deletion flag is false? Is that the intent here?
Can a test be used to confirm the intended behavior here so that we do not accidentally drop databases in case a bug slips through?

xbreizh replied:
Fixed

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+    private Migrator()
+    {
+    }
+    
```
Extra whitespace characters on empty lines

xbreizh replied:
Removed

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+            LOGGER.error("Issue while trying to perform the upgrade.", e);
+        }
+    }
+    
```
Extra whitespace characters on empty lines

xbreizh replied:
Removed

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+    }
+    
```
Extra whitespace characters on empty lines

xbreizh replied:
Removed

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+            dbName = fullConnectionUrl.substring(fullConnectionUrl.lastIndexOf('/') + 1);
+            dbSource.setUsername(username);
+            dbSource.setPassword(password);
+            if(checkDBExists(dbSource) || allowDBDeletion) {
```
Formatting

xbreizh replied:
Adjusted

simranjeetc commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,112 @@
+    
+    private static void createNewDatabase(final String dbName, final BasicDataSource dbSource) throws SQLException
+    {
+            LOGGER.info("\nDB {}- does not exist, or force deletion has been specified for it.\n", dbName);
```
Extra indentation

xbreizh replied:
Removed

simranjeetc commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,107 @@
+            description = "Specifies the name of the database to be created or updated."
+    )
+    private String dbName;
+    
```
Too many blank lines

xbreizh replied:
Removed

simranjeetc commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,107 @@
+
+import picocli.CommandLine;
+
+@CommandLine.Command(mixinStandardHelpOptions = true, name = "util-flywayinstaller")
```
This class has a few empty line with unnecessary whitespace characters.

xbreizh replied:
Removed

dermot-hardy commented on README.md:
```diff
@@ -1 +1,81 @@
+3. Create a Flyway change log called changelog.xml and place this in the `resources` folder of the project.
+4. Build the project with the mainClass as:
+
+   `com.hpe.caf.utilFlywayinstaller.Application`
```
Obviously this shouldn't be `com.hpe`.  Same multiple times below.

xbreizh replied:
Updated to package name

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,109 @@
+        <version>2.5.0-SNAPSHOT</version>
+    </parent>
+    
+    <groupId>com.github.cafapi</groupId>
```
We're trying to move to each repository having its own unique home group id (or a small number of home group ids if that makes more sense).

Discussed in [SCMOD-13857](https://portal.digitalsafe.net/browse/SCMOD-13857) and co. for example.

xbreizh replied:
Updated for com.github.cafapi.util.flywayinstaller

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,109 @@
+    <parent>
+        <groupId>com.github.cafapi</groupId>
+        <artifactId>caf-common-parent</artifactId>
+        <version>2.5.0-SNAPSHOT</version>
```
Why aren't you using the released version of the parent?  As you know referencing a SNAPSHOT version will cause me to raise a release ticket `caf-common-parent` in the future because I will assume that there is some reason you're using it instead of `2.4.0-254`.

xbreizh replied:
Adjusted

dermot-hardy commented on pom.xml:
```diff
@@ -0,0 +1,109 @@
+        <enforceCorrectDependencies>true</enforceCorrectDependencies>
+        <enforceBannedDependencies>true</enforceBannedDependencies>
+        <slf4j.version>1.7.30</slf4j.version>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
```
Aren't these properties inherited from `caf-common-parent`?

xbreizh replied:
Removed

xbreizh commented:
I added a log level toggler to match what we have on the liquibase utility (also, I believe that can be useful).
I removed the fullConnectionString as per above comment (we don't need both and we can then limit some extra check / stripping)

dermot-hardy commented:
What'g going on with this PR?  The Jira is in the Under Review column and yet there is no review requested for this?

dermot-hardy commented on README.md:
```diff
@@ -1 +1,81 @@
+
+1. Create a new Java project.
+2. Add com.github.cafapi.util.flywayinstaller as a dependency.
+3. Create a Flyway change log called changelog.xml and place this in the `resources` folder of the project.
```
Does this seem right to you?  When I see mistakes like this I feel like you're asking me to review work that you haven't reviewed yourself.

xbreizh replied:
Updated

dermot-hardy commented on README.md:
```diff
@@ -1 +1,81 @@
+
+This will produce a .jar specific to your project.
+
+Note: The installer comes prepackaged only with Postgresql and h2 drivers. If you need to use another database driver you must add it as a dependency to your own project.
```
Where is the H2 driver is in this project?

xbreizh replied:
There is none, only postgres. Removed

dermot-hardy commented on README.md:
```diff
@@ -1 +1,81 @@
+*   fd  :  Enables the deletion of the existing database for a fresh install, rather than updating the database.
+*   log : Specifies the logging level of the installer. Valid options are: [debug, info, warning, severe, off]
+
+You can specify these options in a database.properties file. This can be specified with the system property `-DDATABASE_CONFIG=<DIRECTORY OF PROPERTY FILE>`   
```
Is any of this right?  I haven't been able to find `DATABASE_CONFIG` mentioned in the code at all.

xbreizh replied:
Removed

dermot-hardy commented on README.md:
```diff
@@ -1 +1,84 @@
+*   `db.user`  :  Specifies the username to access the database.
+*   `db.pass`  :  Specifies the password to access the database.
+*   `db.connection`  : Specifies the jbdc connection string to the database service. This does not include the database name.  e.g. jdbc:postgresql:/localhost:3307/
+*   `db.connection.url`  : Specifies the jbdc connection string to the database service. This includes the database name.  e.g. jdbc:postgresql:/localhost:3307/storage
```
Wasn't this removed?

xbreizh replied:
Removed from the readme as well

dermot-hardy commented on README.md:
```diff
@@ -1 +1,84 @@
+*   `db.connection.url`  : Specifies the jbdc connection string to the database service. This includes the database name.  e.g. jdbc:postgresql:/localhost:3307/storage
+*   `db.name`  :  Specifies the name of the database to be created or updated.
+*   `fd`  :  Enables the deletion of the existing database for a fresh install, rather than updating the database.
+*   `log` : Specifies the logging level of the installer. Valid options are: [debug, info, warning, severe, off]
```
I don't see `severe` or `off` in the enumeration?  Do you definitely want to have this option, instead of just using the facilities in `caf-logging`?

xbreizh replied:
Updated. I added it to match the liquibase updater, we could remove it but I find it quite useful when using it with a small project

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+            + " password: {}", allowDBDeletion, connectionString, dbName, username, password);
+
+        LOGGER.info("Starting migration ...");
+        try (final BasicDataSource dbSource = new BasicDataSource()) {
```
Why are we using DBCP here?  I can't see the advantage of using connection pooling from this utility.

xbreizh replied:
Using PGSimpleDataSource now which extends BaseDataSource. We could go more generic, but as we are already tied up with postgres, I thought it would be better using it

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static String checkAndConvertConnectionUrl(final String connectionUrl) throws InvalidConnectionStringException
+    {
+        final String connectionUrlWithSlashes = connectionUrl.replace("\\", "/");
```
This seems so strange.  Why are we doing this?

xbreizh replied:
This is to make sure that we have the right url syntax as it url received from windows or linux. That's belt and braces and can be removed if you want

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+    private static String checkAndConvertConnectionUrl(final String connectionUrl) throws InvalidConnectionStringException
+    {
+        final String connectionUrlWithSlashes = connectionUrl.replace("\\", "/");
+        final Matcher matcher = Pattern.compile(CONNECTION_URL_REGEX).matcher(connectionUrlWithSlashes);
```
What is this all about?  What are we trying to achieve here?

xbreizh replied:
Sometimes, we could receive an url like: `jdbc:postgresql://localhost:5437/?adaptiveFetch=false&adaptiveFetchMaximum=-1&adaptiveFetchMinimum=0`
This is to unsure that we get the right format and then to split it to `jdbc:postgresql://localhost:5437/`

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+    {
+        try (final Connection connection = dbSource.getConnection();
+             final Statement statement = connection.createStatement()) {
+            final boolean exists = statement.executeQuery("SELECT * FROM pg_database WHERE datname=lower('" + dbName + "');").next();
```
- What if `dbName` contained a `'`?  This is classic SQL injection.  You should be using a parameterized query.
- Is this definitely a safe way to check if the database exists?  I presume the `pg_database` is cluster-wide and so I'm wondering if super-user permissions are required to check this?  Although I suppose they would be required anyway for `CREATE DATABASE` so it's probably ok.

xbreizh replied:
https://stackoverflow.com/a/26582770/10218144 It seems we can't use the preparedStatement for drop and create as we don't want to have 'dbName'.
If we received a dbName containing a ' like
`CREATE DATABASE 'test';`
then it would fail with:
`SQL Error [42601]: ERROR: syntax error at or near "'test'"`

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+        final String dbName) throws SQLException
+    {
+        final String savedUrl = dbSource.getUrl();
+        dbSource.setUrl(dbSource.getUrl() + "postgres");
```
I don't get it.

xbreizh replied:
For dropping a database, we can't be connected to it, so that would switch back to postgres and allow the operation to be performed. We then would need to switch the url to the targeted database in order to perform the migration.
This is just to explain why I did that back then. Since we don't drop the database anymore, I implemented it differently

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+             final Statement statement = connection.createStatement()) {
+            if (exists) {
+                LOGGER.info("force deletion has been specified.\nDeleting database {}", dbName);
+                statement.executeUpdate("DROP DATABASE " + dbName);
```
Looks like more SQL injection potential to me.

xbreizh replied:
Same as [here](https://github.com/CAFapi/util-flyway-installer/pull/1#discussion_r788097242) We could potentially use a regex for validation the database name using that [documentation](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)

xbreizh replied:
Removed

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+                statement.executeUpdate("DROP DATABASE " + dbName);
+                LOGGER.info("DELETED database: {}", dbName);
+            }
+            statement.executeUpdate("CREATE DATABASE " + dbName);
```
Looks like more SQL injection potential to me.

xbreizh replied:
Same as [here](https://github.com/CAFapi/util-flyway-installer/pull/1#discussion_r788097242) We could potentially use a regex for validation the database name using that [documentation](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)

xbreizh replied:
Removed

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+
+            LOGGER.info("About to perform DB update.");
+            final Flyway flyway = Flyway.configure()
+                .dataSource(dbSource.getUrl() + dbName, username, password)
```
Why aren't we just passing the data source?

xbreizh replied:
now passing the data source

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+            LOGGER.info("About to perform DB update.");
+            final Flyway flyway = Flyway.configure()
+                .dataSource(dbSource.getUrl() + dbName, username, password)
+                .baselineOnMigrate(true)
```
Why are we setting this?

xbreizh replied:
This will set the first Flyway migration if not existent. It will initialize the schema history table. https://flywaydb.org/documentation/configuration/parameters/baselineOnMigrate

dermot-hardy replied:
I read the link you provided but I still don't see how it makes sense to set this if we're not setting `baselineVersion`?

xbreizh replied:
This is to avoid that exception that would occur if the database is not empty: 

`[12:03:52.104Z #6e8.001 ERROR -            -   ] c.g.c.u.f.Migrator: {"message":"Found non-empty schema(s) \"public\" but no schema history table. Use baseline() or set baselineOnMigrate to true to initialize the schema history table."`

We don't have any baselineVersion to refer to at the moment as it will be our first migration. We could potentially add another argument later on to specify it.

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,117 @@
+                .baselineOnMigrate(true)
+                .load();
+            flyway.migrate();
+            flyway.validate();
```
What is the purpose of this?  Wouldn't the `migrate()` have thrown an exception if there was an issue?  Is it just belt and braces or is it something else?

xbreizh replied:
Did test it earlier with keeping only migrate and also removing a rown from flyway_history table and getting `[12:43:01.419Z #5b4.001 ERROR -            -   ] c.g.c.u.f.Migrator: {"message":"Validate failed: Migrations have failed validation\nDetected resolved migration not applied to database: 2. To ignore this migration, set -ignoreIgnoredMigrations=true. To allow executing this migration, set -outOfOrder=true.\nNeed more flexibility with validation rules? Learn more: https://rd.gt/3AbJUZE","exception":"org.flywaydb.core.api.exception.FlywayValidateException: Validate failed: Migrations have failed validation\nDetected resolved migration not applied to database: 2. To ignore this migration, set -ignoreIgnoredMigrations=true. To allow executing this migration, set -outOfOrder=true.\nNeed more flexibility with validation rules? Learn more: https://rd.gt/3AbJUZE\r\n\tat org.flywaydb.core.Flyway$1.execute(Flyway.java:131)\r\n\tat org.flywaydb.core.Flyway$1.execute(Flyway.java:124)\r\n\tat org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:214)\r\n\tat org.flywaydb.core.Flyway.migrate(Flyway.java:124)\r\n\tat com.github.cafapi.util.flywayinstaller.Migrator.migrate(Migrator.java:75)\r\n\tat com.github.cafapi.util.flywayinstaller.Application.call(Application.java:96)\r\n\tat com.github.cafapi.util.flywayinstaller.Application.call(Application.java:27)\r\n\tat picocli.CommandLine.executeUserObject(CommandLine.java:1953)\r\n\tat picocli.CommandLine.access$1300(CommandLine.java:145)\r\n\tat picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)\r\n\tat picocli.CommandLine$RunLast.handle(CommandLine.java:2352)\r\n\tat picocli.CommandLine$RunLast.handle(CommandLine.java:2314)\r\n\tat picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)\r\n\tat picocli.CommandLine$RunLast.execute(CommandLine.java:2316)\r\n\tat picocli.CommandLine.execute(CommandLine.java:2078)\r\n\tat com.github.cafapi.util.flywayinstaller.Application.main(Application.java:84)\r\n"}` 

So indeed the migrate does also perform validation. There was surely a reason why I added the validate on top of it back then but I can't recall so removing it

dermot-hardy commented on README.md:
```diff
@@ -1 +1,79 @@
+
+1. Create a new Java project.
+2. Add `com.github.cafapi.util.flywayinstaller` as a dependency.
+3. Create a folder called `db.migration` in the `resources` folder of the project.
```
Is it a folder called `db.migration` or a `migration` folder inside a `db` folder inside `resources`?  Or is either fine?

xbreizh replied:
either is fine. I updated it with what is declared in the [Flyway documentation](https://flywaydb.org/documentation/getstarted/firststeps/maven)

> Create the migration directory src/main/resources/db/migration.

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,138 @@
+    private static final String CONNECTION_URL_REGEX = "^.*\\/\\/.+:\\d+\\/";
+    private static final String CREATE_DATABASE = "CREATE DATABASE \"%s\"";
+    private static final String DOES_DATABASE_EXIST
+        = "SELECT EXISTS ( SELECT datname FROM pg_catalog.pg_database WHERE lower( datname ) = lower( ? ) );";
```
You're only checking row existence so you don't need to output the column.
```suggestion
        = "SELECT EXISTS (SELECT NULL FROM pg_catalog.pg_database WHERE lower( datname ) = lower( ? ));";
```

xbreizh replied:
Done

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,138 @@
+    {
+        try (final Connection connection = dbSource.getConnection();
+             final Statement statement = connection.createStatement();) {
+            statement.executeUpdate(String.format(CREATE_DATABASE, dbName));
```
Are you sure there isn't a way to use a prepared statement for this?  I feel like there should be.

xbreizh replied:
Updated to use prepared statements

dermot-hardy replied:
You're following the letter of the law but not the spirit (i.e. the sql is still being dynamically put together with concatenation - this is what the advice to use prepared statements is meant to avoid).

dermot-hardy replied:
Reading the documentation I don't see a straight forward way of doing this.  The best thing is probably to execute this as a prepared statement:
```
SELECT format('CREATE DATABASE %I', ?)
```

and then to retrieve and execute the result.

xbreizh replied:
Nice one, updated.

> Why have you dropped the use of `caf-logging`?
> 
> [30b81aa#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L91-L95](https://github.com/CAFapi/util-flyway-installer/commit/30b81aab736c8d35071481cbac464231c1489ec7#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L91-L95)

@dermot-hardy The initial idea here was to remove the jackson dependency (included in the caf-logging) as it's not being used and as it was bringing conflicts in the job-service. The flyway migrator being used so far solely in the job service, and the job-service already using caf-logging, I thought it would be fine to exclude it from here. But I understand that the utility is supposed to be usable on its own and if so, would expose the logs to possible threats (and also to harmonize log formatting), so I restored it and excluded the jackson dependency from it (317kb).
I also excluded logback-core from it as it was conflicting with the newer version used for compiling: 1.2.10

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,138 @@
+                resetOrCreateDatabase(dbSource, dbName);
+            }
+            // Update dbSource url to include the database name
+            dbSource.setUrl(urlAfterConversion + dbName);
```
Can't you use [`.setDatabaseName()`](https://jdbc.postgresql.org/documentation/publicapi/org/postgresql/ds/common/BaseDataSource.html#setDatabaseName-java.lang.String-)?

xbreizh replied:
Done + no longer using url but `server, port, dbName, user and pwd`

dermot-hardy commented on util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java:
```diff
@@ -0,0 +1,138 @@
+            flyway.migrate();
+            LOGGER.info("DB update finished.");
+
+        } catch (final Exception e) {
```
- What are you trying to catch here?  The method is declared to throw `InvalidConnectionStringException` and `SQLException` but if you're doing a generic catch-all here then how will they ever be thrown?
- Why are we trying to catch anything here anyway?  Wouldn't we want the exception to be thrown back to the CLI code for it to log and return the appropriate error codes?

xbreizh replied:
removed the try catch so we propagate the exception.

dermot-hardy reviewed (changes requested):
Why have you dropped the use of `caf-logging`?

https://github.com/CAFapi/util-flyway-installer/pull/1/commits/30b81aab736c8d35071481cbac464231c1489ec7#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L91-L95

dermot-hardy commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,121 @@
+        } catch (final SQLException ex) {
+            LOGGER.error("Issue while trying to perform the migration.", ex);
+            return 1;
+        } catch (final Exception ex) {
```
What type of exception are you trying to catch here?

xbreizh replied:
2 types of exceptions mainly: authentication, then anything else that would classify as migration issues. I adapted the code to reflect it. Picocli takes already care of the invalid input exceptions

dermot-hardy commented on util-flywayinstaller-cli/src/main/java/com/github/cafapi/util/flywayinstaller/Application.java:
```diff
@@ -0,0 +1,121 @@
+            Migrator.migrate(dbHost, dbPort, dbName, username, password);
+        } catch (final SQLException ex) {
+            LOGGER.error("Issue while trying to perform the migration.", ex);
+            return 1;
```
You've adjusted this but you didn't adjust the `exitCodeList` above to match.

xbreizh replied:
Done
